text/0813-template-layering.md

```diff
@@ -0,0 +1,332 @@
+  as a standalone expression), providing the same high-level programming model
+  for authoring components in standard JavaScript environments.
+
+- It requires manually supplying the lexical scope variable bindings.
+
+- While it is more cumbersome to use and have a degraded experience (e.g.
+  possibly lacking hbs syntax highlighting) compared to first-class component
+  templates, it is still designed to be ergonomic to use, to the extent
+  possible in standard JavaScript environments.
+
+- It is designed to be pre-processed at build time where possible, just like
+  the `<template>` feature and today's "official" `hbs` tag. However, in cases
+  where this is not possible, there will be a runtime implementation available
+  provided the template compiler is also available at runtime.
+
+- For environments where it is possible to configure the development and build
+  tools to recognize the format, the tagged template literal form can be used,
+  which automatically captures the lexical scope variable bindings. This
+  directly replaces [ember-template-imports][ember-template-imports]'s "hbs
+  backtick" format but brings its naming into alignment with the `<template>`
+  feature, and uses a static initializer block to associate the template.
+
+  This form will produce an error at runtime if not pre-processed out
+  as a correct runtime implementation is impossible.
+
+  Despite the drawbacks mentioned above, this format may still be preferred as
+  a transitional tool for early adopters while editor integrations are being
+  worked on. It may also be useful for communicating with code snippets in
+  platforms where syntax highlighting is not yet available for the custom
+  `.gjs` format (such as Discord and GitHub today), but where it's desirable to
+  still retain the highlighting for the JavaScript/TypeScript portions.
+
+## Detailed design
+
+This RFC proposes to re-specify the `<template>` language extension into a
+de-sugaring into `template()` calls:
+
+1. Top-level declaration
+
+   ```
+   import Foo form 'somewhere';
+
+   <template>
+     <Foo />
+   <template>
+   ```
+
+   becomes
+
+   ```js
+   import { template } from '@ember/template-compilation';
+   import Foo form 'somewhere';
+
+   export default template(`
+     <Foo />
+   `, () => ({ Foo }));
+   ```
+
+2. Expression
+
+   ```
+   import Foo form 'somewhere';
+
+   export const Bar = <template><Foo /><template>;
+   ```
+
+   becomes
+
+   ```js
+   import { template } from '@ember/template-compilation';
+   import Foo form 'somewhere';
+
+   export const Bar = template(`<Foo />`, () => ({ Foo }));
+   ```
+
+3. Class
+
+   ```
+   import Foo form 'somewhere';
+
+   export default class Bar {
+     <template><Foo /><template>
+   }
+   ```
+
+   becomes
+
+   ```js
+   import { template } from '@ember/template-compilation';
+   import Foo form 'somewhere';
+
+   export default @template(`<Foo />`, () => ({ Foo }) class Bar {
```

Comment:
can / should this RFC mention the pre-decorators version of class decorator usage?

I think it'd be:

```js
export default template(`<Foo />`, () => ({ Foo }))(class Bar {
});
```

Reply:
Yeah we can definitely bikeshed what is the most ergonomic version of this should be. I don't super love the decorator approach either, I just pushed a commit for a version using a static initializer block (matching the template/backtick version)

Reply:
well, my motivation here is, I think aligned with the motivation of this RFC, to potentially allow shipping of ember "stuff" without transpilation.  Who knows how long it'll take before decorators are native, ya know?

Reply:
What do you think about the static initializer block version? I like that it keeps it inside the class body, matching `<template>` more closely. It always does not _intrinsically_ require the use of static initializer blocks if it's not supported in your environment, it's basically just a "three arguments version" of the function and you can wrap it around the class if you want to.

Reply:
I pushed an update making the static initializer block version the preferred one, and moved the class decorator to the alternatives section.

Reply:
I have internal conflict. So like,

I'm used to having to do:
```js
setComponetnTemplate(hbs`my stuff`, MyClass);
```
which _could_ look like very similar to the decorator usage of template (which, maybe for other readers, class decorators are syntax sugar for functions that take a class and return a class)
```
@setComponentTemplate(hbs`my stuff`)
class MyClass {

}
```

<details><summary>implementation</summary>

```js
export function setComponentTemplate(...args) {
  if (args.length === 2) {
    _setComponentTemplate(...args);
    
    return args[1];
  }
  
  if (args.length === 1) {
    return (Klass) => {
      _setComponentTemplate(args[0], Klass);
      
      return Klass;
    }
  }
  
  throw new Error(`Unsupported arity: ${args.length}`);
}

```

</details>

Now, for the static initializer block -- I had never seen these in use before now, so that's exciting (TIL, etc) -- I can't _comfortably_ have an informed opinion on them yet though, only gut reactions.

And my gut reaction is that the static initialize block achieves the intent we want:
 - a component's template is not dynamic (or rather, you're in for a world of pain if you make it dynamic)
 - it's clear that the template is the same for every instance of the component
 - the association between template and component is also clear (whereas in can be missed with things like `setComponentTemplate`
 
 And what I'm wary of:
  - people unfamiliar with static initializers may be like "wtf? what is `this` in a static context", "why is JS weird?" (lol), "not `this` again" -- to which I say, _learn the language you're using, ask questions, etc,_
  - it's more typing / more blocks than just calling the function (but with the function, you have the same problem as `setComponentTemplate`, where the association is less obvious.
  
  I'm actually just realizing that we can use static initializers with `setComponentTemplate` today... which I think I may want to explore in my own stuff:
  
  ```js
  class Demo extends Component {
    static {
      setComponentTemplate(hbs`hello @name`, this);
    }
  }
  ```


So.... I think I'm in _favor_ of static initializers _because_ of the implied association between template and component.